pkg/common/jwtsvid/simple.go

```diff
@@ -0,0 +1,162 @@
+package jwtsvid
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"time"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/spiffe/spire/pkg/common/cryptoutil"
+	"github.com/spiffe/spire/pkg/common/idutil"
+)
+
+const (
+	certificateThumbprintHeader = "x5t#S256"
+)
+
+func SignSimpleToken(spiffeID string, audience []string, expires time.Time, signer crypto.Signer, cert *x509.Certificate) (string, error) {
+	if err := validateKeypair(signer, cert); err != nil {
+		return "", err
+	}
+
+	if err := idutil.ValidateSpiffeID(spiffeID, idutil.AllowAnyTrustDomainWorkload()); err != nil {
+		return "", err
+	}
+
+	// cap expiration to the signing certificate expiration, if necessary.
+	if expires.IsZero() {
+		return "", errors.New("expiration is required")
+	}
+	if !cert.NotAfter.IsZero() && expires.After(cert.NotAfter) {
+		expires = cert.NotAfter
+	}
+
+	if len(audience) == 0 {
+		return "", errors.New("audience is required")
+	}
+
+	claims := jwt.MapClaims{
+		"sub": spiffeID,
+		"exp": expires.Unix(),
+		"aud": audienceClaim(audience),
+	}
+
+	token := jwt.NewWithClaims(signingMethodES256, claims)
+	token.Header[certificateThumbprintHeader] = certificateThumbprint(cert)
+	signedToken, err := token.SignedString(signer)
+	if err != nil {
+		return "", err
+	}
+	return signedToken, nil
+}
+
+type SimpleTrustBundle interface {
+	TrustDomain() string
+	FindCertificate(ctx context.Context, thumbprint string) (*x509.Certificate, error)
+}
+
+type simpleTrustBundle struct {
+	trustDomain string
+	lookup      map[string]*x509.Certificate
+}
+
+func NewSimpleTrustBundle(trustDomain string, certs []*x509.Certificate) SimpleTrustBundle {
+	t := &simpleTrustBundle{
+		trustDomain: trustDomain,
+		lookup:      make(map[string]*x509.Certificate),
+	}
+	for _, cert := range certs {
+		t.lookup[certificateThumbprint(cert)] = cert
+	}
+	return t
+}
+
+func (t *simpleTrustBundle) TrustDomain() string {
+	return t.trustDomain
+}
+
+func (t *simpleTrustBundle) FindCertificate(ctx context.Context, thumbprint string) (*x509.Certificate, error) {
+	cert := t.lookup[thumbprint]
+	if cert == nil {
+		return nil, errors.New("signing certificate not found in trust bundle")
+	}
+	return cert, nil
+}
+
+func ValidateSimpleToken(ctx context.Context, token string, trustBundle SimpleTrustBundle, audience string) (jwt.MapClaims, error) {
+	claims := make(jwt.MapClaims)
+	if _, err := jwt.ParseWithClaims(token, claims, func(t *jwt.Token) (interface{}, error) {
```

Comment:
Hmm, because of the amount of captured variables the code uses (i.e. ctx, trustBundle, claims), it actually is a bit gross to move it out.... If you're ok with it I'm gonna keep it as is.